### ISSUE TITLE
Bond objects retrieved from particles don't match original

### ISSUE THREAD

**Issue by RudolfWeeber** [espressomd/espresso#4391]
MNWE:
```
import espressomd
from espressomd.interactions import HarmonicBond

s=espressomd.System(box_l=[1]*3)
h=HarmonicBond(k=1,r_0=0)
s.bonded_inter.add(h)
p1=s.part.add(pos=[0]*3)
p2=s.part.add(pos=[0]*3)
b=[(h,p2.id)]
p1.bonds=b
print(p1.bonds)
print(b)
assert p1.bonds == b # which fails
```

I susepct the reason is that a new ScriptObject is instanced when retrieving a bond by id

```
    def __getitem__(self, bond_id):
        return self._get_bond(bond_id)

    def _get_bond(self, bond_id):
[...]
        # Find out the type of the interaction from ESPResSo
        bond_type = get_bonded_interaction_type_from_es_core(bond_id)
[...]
        # Find the appropriate class representing such a bond
        bond_class = bonded_interaction_classes[bond_type]

        # Create a new script interface object (i.e. a copy of the shared_ptr)
        # which links to the bonded interaction object
        return bond_class(bond_id)
```

I think that, whil it represents the sem bond, it creates a new script object.
Wouldn't it be possible to return the existing bond script object from the ObjectMap (via call-method). See the implementation in the breaking_bonds branch on my github repo.

**Comment by jngrad** [comment]
It does create a new script object, but we adapted the `__richcmp__()` method such that we compare the sip. Bond objects are perfectly comparable, and we have a test for it:
https://github.com/espressomd/espresso/blob/b801493a3390b0d03d03bfe4306161f404648d0f/testsuite/python/interactions_bonded_interface.py#L121-L123
Your MNWE fails because you're comparing a tuple to a list, which must evaluate to False. Change the last line to this:
```python
assert p1.bonds == tuple(b)
```

**Comment by RudolfWeeber** [comment]
Ah, missed that. Sorry for the noise.

That aside, what was the reason for instancing a new ScriptObject rather than returning the existing one stored by the ObjectMap?

**Comment by jngrad** [comment]
> That aside, what was the reason for instancing a new ScriptObject rather than returning the existing one stored by the ObjectMap?

I can have a look.

**Comment by jngrad** [comment]
> > That aside, what was the reason for instancing a new ScriptObject rather than returning the existing one stored by the ObjectMap?
> 
> I can have a look.

We initially chose this approach because that's how the constraints framework was designed. The advantage is that you only have one hashmap in the core, instead of maintaining a second hashmap in the script interface that must always match the one in the core. It's tricky, because there are situations where we need to make the SI hashmap disagree with the core hashmap:
* When we reload from a checkpoint, the SI hashmap is only restored on the head node, hence the other ranks disagree with the core. We ended up disabling deserialization of the SI hashmap on the head rank to avoid MPI ranks disagreeing with each other. When reloading from a checkpoint, we repopulate the SI hashmaps on all ranks by creating new SI bonds from their parameter dicts.
* The particle list and the collision detection both need to retrieve bond objects. This is achieved by instantiating a new temporary (and empty!) SI hashmap to retrieve one SI bond indirectly via the core and construct a new SI bond from it, using  `bond = BondedInteractions()[bond_id]`.

There a two major obstacles to refactoring that code. First, passing the `system.bonded_inter` object as a reference to the `ParticleList`, `ParticleSlice`, `ParticleHandle`, `CollisionDetection` and their helper free functions is really difficult; I tried on Tuesday and couldn't get it to work. It's easier to instead store a reference to `system.bonded_inter` in a private global of the `particle_data.pyx` and `collision_detection.pyx` modules. The second issue is a deal-breaker: we have no control over the order in which the client classes are reloaded from the checkpoint. Since the `ParticleList` needs bond information to deserialize its particles, the `BondedInteractions` must be available beforehand. This just so happened to be the case until now. I'm not sure which factors play a role in the deserialization order, in any case right now `BondedInteractions` and `ParticleList` are deserialized before `System`, so we cannot pass the SI hashmap as reference to client classes or to private module globals. The current order is as follows:
```
IntegratorHandle::__setstate__
BondedInteractions::__setstate__
ParticleList::__setstate__
Actors::__setstate__
Thermostat::__setstate__
CollisionDetection::__setstate__
System::__setstate__
  _globals
  integrator
  max_oif_objects
  _active_virtual_sites_handle
  non_bonded_inter
  bonded_inter
  cell_system
  part
  actors
  analysis
  auto_update_accumulators
  comfixed
  constraints
  galilei
  lbboundaries
  thermostat
  collision_detection
CellSystem::__setstate__
```

I will adapt the SI hashmap such that it returns the SI object directly if it exists, otherwise the old implementation kicks in and creates a new SI bond object for it. If we find a solution in the future to control the order of deserialization or if we remove the `ParticleList` and `CollisionDetection` dependency on the SI bonds hashmap, then we can remove the `else` branch in `BondedInteractions._get_bond()`.

While looking into the script interface, I also found that the `BondedInteractions` object was not properly deserialized, which makes it impossible to call `BondedInteractions.call_method()` without causing a segfault. This regression wasn't caught in the checkpointing tests. There is also this regression:
https://github.com/espressomd/espresso/blob/b801493a3390b0d03d03bfe4306161f404648d0f/src/python/espressomd/interactions.pyx#L1732-L1738
As we learned recently in the walberla branch, if we have `@property` decorators in a class derived from `ScriptInterfaceHelper`, the corresponding property setters will never execute. I've now found the root cause: `ScriptInterfaceHelper.__setattr__()` is broken and doesn't look up property setters. I've fixed it (fa2d52b629b8d20e) and removed the workaround from the walberla branch.

I'll open a PR to fix this on the python branch.
